### PR TITLE
ParticleTracker refactor: Bring in particle_integrators.py

### PR DESCRIPTION
**Pull request by StanczakDominik**
This is (finally!) a follow-up to the closed #675. It brings in a separated implementation of the Boris algorithm, which I was hoping would solve the issue raised in https://plasmapy.discourse.group/t/e-cross-b-drift/; we're not there yet, but this is a change in the right direction and should be easier to review.

7f20252 is marked [REVERT], but I haven't yet found a way to combine astropy.units with parallel numba, and in my previious tests the parallelization did not bring huge speedups, so it's fine to keep it that way for now.

This separate implementation will also make #895 easier.

### CODE REVIEW

**Comment by StanczakDominik** [comment]
`black`'s failure gets fixed in #952.

**Comment by pheuer** [comment]
This PR would also address a couple of the points raised in issue #936

**Comment by StanczakDominik** [comment]
Merging this now would break Python 3.9 support, because that version broke Numba (https://github.com/numba/numba/issues/6345) and we won't get that fixed until RC 2021.

**Comment by rocco8773** [comment]
Could you possibly have a `numba` and non-`numba` version by doing something like....

```python

try:
    import numba
    _range = numba.prange
except ImportError:
    warn(
        "If running Python < 3.9 consider installing numba.  This will allow parallel "
        "computing and speed up ParticleTracker"
    )
    _range = np.arange

def foo(*args):
    # use _range for the loops so the function is automatically a numba or 
    # non-numba version
    ...

if "numba" in globals:
    foo = numba.njit()(foo)
```

Then we might be able to use [environment markers](https://www.python.org/dev/peps/pep-0508/#environment-markers) so `numba` is only installed on `python < 3.9`. (see also https://stackoverflow.com/a/32643122)  I've never done this, so I'm not sure how well it will work.

**Comment by StanczakDominik** [comment]
It's still a for loop. If we do "optional Numba" here, the performance will suffer greatly over the current implementation.

**Comment by rocco8773** [comment]
> It's still a for loop. If we do "optional Numba" here, the performance will suffer greatly over the current implementation.

I get that.  The idea is that, with the environmental markers we can make `numba` a dependency for installs on `python < 3.9` and thus utilized the `numba` functionality you're integrating here.  If someone is using `python >= 3.9`, then `numba` will not be installed, the particle tracker functionality will revert to a slower form, and the user is warned about what is happening.  However, the UI is made consistent across the python versions so when the `numba` with `python39` issue is resolved the user does not have to change anything when make the patch on our end.

**Comment by StanczakDominik** [comment]
I went without numba for now. I figure we can wait until Q1 2021 and introduce less complexity.

In the meantime, I went with refactoring the numpy version out of the class, and lo and behold, I fixed the ExB drift direction bug. I've got the numba version backed up, but for now, this should be good.

**Comment by StanczakDominik** [comment]
Looks ready to review! MacOS build cancellation smells of intermittency.

File: changelog/953.bugfix.rst
Comment (namurphy) [inline changelog/953.bugfix.rst:2]:
```suggestion
error, thus the E×B drift came out in the incorrect direction.
```
Unicode!

File: plasmapy/simulation/particle_integrators.py
Comment (namurphy) [inline plasmapy/simulation/particle_integrators.py:20]: I'm wondering about adding some more detail to the arguments.  Reading this, I could end up being confused about the `ndarray` objects...like if they are three-vectors corresponding to x, y, z.  I could also get confused about what units are assumed.

File: plasmapy/simulation/particle_integrators.py
Comment (namurphy) [inline plasmapy/simulation/particle_integrators.py:61]: Huh, I can't remember seeing this notation before.  I've pretty much always done `v[:] =`.  Interesting to learn!

**Review by namurphy — CHANGES_REQUESTED** [review]
Looking good to me so far!  My main comment is to add a little bit more detail in the `boris_push` docstring to reduce the chance of ambiguity.  Thanks!

**Comment by pheuer** [comment]
@StanczakDominik As expected I found another sign error elsewhere in my proton radiography code, so I'm now pretty sure you have the signs right in this version of `boris_push` (eg. no minus sign in the definition of `t`).

File: changelog/953.bugfix.rst
Comment (StanczakDominik) [reply changelog/953.bugfix.rst:2]: That's some good application of unicode, I love it! But I think it's confusing enough for the mobile GitHub app that I can't merge the suggestion without a crash... Will merge after reporting it:D

File: plasmapy/simulation/particle_integrators.py
Comment (StanczakDominik) [reply plasmapy/simulation/particle_integrators.py:61]: Yeah, that would have been fine as well! In fact it might have some subtle differences, I'll check back on that - thanks for pointing it out...

File: plasmapy/simulation/particle_integrators.py
Comment (pheuer) [inline plasmapy/simulation/particle_integrators.py:69]: The Lorentz factor is already implemented in `formulary.relativity' (although that version is for velocities). Maybe it would be better to modify that version to accept momentum as well (calculating gamma as appropriate) and then import that here?

Comment (StanczakDominik) [reply plasmapy/simulation/particle_integrators.py:69]: Fair point, though I did want to have a version without the overhead on units in here. For now, it's not being used (I had another pusher in here that did incorporate it) so I'll just remove it for now.

**Comment by StanczakDominik** [comment]
Should be good to go!

**Comment by StanczakDominik** [comment]
@namurphy here's the promised reminder :)

File: plasmapy/simulation/particle_integrators.py
Comment (namurphy) [inline plasmapy/simulation/particle_integrators.py:59]: After working on a pull request on physical types in Astropy (https://github.com/astropy/astropy/pull/11204), it's really bugging me to not be able to figure out what sort of physical type corresponds to units of C · s / m!

**Review by namurphy — APPROVED** [review]
I like this refactoring, so thank you!  My final suggestions are pretty minor and mostly aren't even necessary.  I think it's ready to be merged whenever you're ready to.

File: plasmapy/simulation/particle_integrators.py
Comment (StanczakDominik) [reply plasmapy/simulation/particle_integrators.py:59]: I'll have to take a closer look at that one!

**Pull request merged**
